src/hashtree.erl

```diff
@@ -355,41 +368,104 @@ update_tree(State) ->
+
+maybe_clear_buckets(?ALL_SEGMENTS, State) ->
+    Segments = lists:seq(0, State#state.segments div State#state.width - 1),
+    %% Limit the clear to the disk buckets, then dump the whole tree
+    State2 = clear_bucket(State#state.mem_levels, State#state.levels, Segments, State),
+    State2#state{tree = dict:new()};
+maybe_clear_buckets(_Segments, State) ->
+    State.
+
+clear_bucket(MinLevel, Level, _Segments, State) when Level =< MinLevel ->
+    State;
+clear_bucket(MinLevel, Level, Segments, State = #state{width = Width}) ->
+    State2 = lists:foldl(fun(Segment, State1) ->
+                                 del_disk_bucket(Level, Segment, State1)
+                         end, State, Segments),
+    ParentSegments = lists:usort([Segment div Width || Segment <- Segments]),
+    clear_bucket(MinLevel, Level - 1, ParentSegments, State2).
+
 
 -spec update_tree([integer()], hashtree()) -> hashtree().
 update_tree([], State) ->
     State;
-update_tree(Segments, State) ->
+update_tree(Segments, State=#state{next_rebuild=NextRebuild, width=Width,
+                                   levels=Levels}) ->
+    LastLevel = Levels,
     Hashes = orddict:from_list(hashes(State, Segments)),
-    Groups = group(Hashes, State#state.width),
-    LastLevel = State#state.levels,
-    NewState = update_levels(LastLevel, Groups, State),
-    NewState.
+    Groups = group(Hashes, Width),
+    update_levels(LastLevel, Groups, State, NextRebuild).
 
 -spec rehash_tree(hashtree()) -> hashtree().
 rehash_tree(State) ->
-    State2 = snapshot(State),
-    rehash_perform(State2).
+    State2 = flush_buffer(State),
+    State3 = snapshot(State2),
+    rehash_perform(State3).
 
 -spec rehash_perform(hashtree()) -> hashtree().
 rehash_perform(State) ->
-    Hashes = orddict:from_list(hashes(State, ['*', '*'])),
+    Hashes = orddict:from_list(hashes(State, ?ALL_SEGMENTS)),
     case Hashes of
         [] ->
             State;
         _ ->
             Groups = group(Hashes, State#state.width),
             LastLevel = State#state.levels,
-            NewState = update_levels(LastLevel, Groups, State),
+            %% Always do a full rebuild on rehash
+            NewState = update_levels(LastLevel, Groups, State, full),
             NewState
     end.
 
+%% @doc Check if shutdown/closing of tree-id was clean/dirty by comparing
+%%      `closed' to `opened' metadata count for the hashtree, and,
+%%      increment opened count for hashtree-id.
+%%
+%%
+%%      If it was a clean shutdown, set `next_rebuild' to be an incremental one.
+%%      Otherwise, if it was a dirty shutdown, set `next_rebuild', instead,
+%%      to be a full one.
+-spec mark_open_and_check(index_n()|binary(), hashtree()) -> hashtree().
+mark_open_and_check(TreeId, State) when is_binary(TreeId) ->
+    MetaTerm = read_meta_term(TreeId, [], State),
+    OpenedCnt = proplists:get_value(opened, MetaTerm, 0),
+    ClosedCnt = proplists:get_value(closed, MetaTerm, 0),
```

Comment:
Thinking about upgrading, it would be safer to default closed to -1 so that it forced a full rebuild. Existing AAE trees may have had their dirty segments lost on restart.

Anything creating a new known-empty hashtree should explicitly set the open metadata to a known 0,0 before calling mark_open_and_check - maybe mark_open_empty?

Otherwise trees in potentially unknown state are marked as good until the next rebuild is triggered in 7 days.